File: tools/observability/event_log.py
#!/usr/bin/env python3
"""
DuckDB Event Log Service
========================
Billing-grade event log for API calls with cost, status, latency, and run_id tracking.
"""

import argparse
import json
import sys
import uuid
from datetime import datetime
from pathlib import Path

try:
    import duckdb
except ImportError:
    print("ERROR: duckdb package not installed. Run: pip install duckdb", file=sys.stderr)
    sys.exit(1)


def init_database(db_path: str) -> dict:
    """Initialize DuckDB database and schema"""
    try:
            CREATE TABLE IF NOT EXISTS api_event_log (
                id TEXT PRIMARY KEY,
                timestamp TIMESTAMP NOT NULL,
                api_name TEXT NOT NULL,
                endpoint TEXT NOT NULL,
                method TEXT NOT NULL DEFAULT 'GET',
                status_code INTEGER,
                success BOOLEAN NOT NULL,
                latency_ms DOUBLE NOT NULL,
                credits_cost DOUBLE NOT NULL DEFAULT 0,
                run_id TEXT,
                error_message TEXT,
                metadata_json TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );
        """)
        
        con.execute("""
            CREATE INDEX IF NOT EXISTS idx_event_log_timestamp 
            ON api_event_log(timestamp);
        """)
        
        con.execute("""
            CREATE INDEX IF NOT EXISTS idx_event_log_api_name 
            ON api_event_log(api_name);
        """)
        
        con.execute("""
            CREATE INDEX IF NOT EXISTS idx_event_log_run_id 
            ON api_event_log(run_id);
        """)
        
        con.execute("""
            CREATE INDEX IF NOT EXISTS idx_event_log_success 
            ON api_event_log(success);
        """)
        
        return {"success": True}
    except Exception as e:
        return {"success": False, "error": str(e)}


def log_event(db_path: str, event: dict) -> dict:
    """Log an API call event"""
    try:
        
        return {"success": True}
    except Exception as e:
        return {"success": False, "error": str(e)}


def get_credits_in_window(db_path: str, window_start: str) -> dict:
    """Get credits spent in time window"""
    try:
        
        return {"total_credits": total_credits}
    except Exception as e:
        return {"total_credits": 0.0, "error": str(e)}


def get_stats(db_path: str, api_name: str = None, start_date: str = None, end_date: str = None) -> dict:
    """Get event log statistics"""
            "error_count": int(result[4]) if result else 0,
        }
    except Exception as e:
        return {
            "total_events": 0,
            "total_credits": 0.0,
            "success_rate": 0.0,
            "avg_latency": 0.0,
            "error_count": 0,
            "error": str(e),
        }


def main():
    parser = argparse.ArgumentParser(description="DuckDB Event Log Service")
    parser.add_argument("--operation", required=True, choices=["init", "log", "get_credits_in_window", "get_stats"])
    parser.add_argument("--db-path", required=True, help="Path to DuckDB database file")
    parser.add_argument("--event", help="Event data (JSON string)")
    parser.add_argument("--window-start", help="Window start timestamp (ISO format)")
    parser.add_argument("--api-name", help="API name for filtering")
    parser.add_argument("--start-date", help="Start date (ISO format)")
    parser.add_argument("--end-date", help="End date (ISO format)")
    
    args = parser.parse_args()
    
    # Ensure database directory exists
    db_path_obj = Path(args.db_path)
    db_path_obj.parent.mkdir(parents=True, exist_ok=True)
    
    result = {}
    
    if args.operation == "init":
        result = init_database(args.db_path)
    elif args.operation == "log":
        if not args.event:
            result = {"success": False, "error": "Event data required for log operation"}
        else:
            try:
                event_data = json.loads(args.event)
                result = log_event(args.db_path, event_data)
            except json.JSONDecodeError as e:
                result = {"success": False, "error": f"Invalid JSON: {str(e)}"}
    elif args.operation == "get_credits_in_window":
        if not args.window_start:
            result = {"total_credits": 0.0, "error": "Window start required"}
        else:
            result = get_credits_in_window(args.db_path, args.window_start)
    elif args.operation == "get_stats":
        result = get_stats(args.db_path, args.api_name, args.start_date, args.end_date)
    
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: packages/observability/src/event-log.ts
/**
 * DuckDB Event Log Service
 * ========================
 * Billing-grade event log for API calls with cost, status, latency, and run_id tracking.
 * Uses DuckDB for efficient storage and querying via Python engine.
 */

import { logger, PythonEngine, AppError, ConfigurationError } from '@quantbot/infra/utils';
import { z } from 'zod';
import { DateTime } from 'luxon';
import { join } from 'path';

/**
 * Event log entry schema
 */
export const EventLogEntrySchema = z.object({
  id: z.string().uuid().optional(),
  timestamp: z.date(),
  api_name: z.string(),
  endpoint: z.string(),
  method: z.string().default('GET'),
  status_code: z.number().nullable(),
  success: z.boolean(),
  latency_ms: z.number(),
  credits_cost: z.number().default(0),
  run_id: z.string().optional(),
  error_message: z.string().nullable().optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
});

export type EventLogEntry = z.infer<typeof EventLogEntrySchema>;

/**
 * Circuit breaker configuration
 */
export interface CircuitBreakerConfig {
  /**
   * Time window in minutes to check credits
   */
  windowMinutes: number;
  /**
   * Maximum credits allowed in the window
   */
  threshold: number;
  /**
   * Whether circuit breaker is enabled
   */
  enabled: boolean;
}

/**
 * Event log service configuration
 */
export interface EventLogConfig {
  /**
   * Path to DuckDB database file
   */
  dbPath: string;
  /**
   * Circuit breaker configuration
   */
  circuitBreaker?: CircuitBreakerConfig;
}

/**
 * Circuit breaker state
 */
interface CircuitBreakerState {
  tripped: boolean;
  trippedAt: Date | null;
  lastCheck: Date;
}

/**
 * DuckDB Event Log Service
 */
export class EventLogService {
  private pythonEngine: PythonEngine;
  private dbPath: string;
  private circuitBreaker: CircuitBreakerConfig;
  private circuitState: CircuitBreakerState = {
    tripped: false,
    trippedAt: null,
    lastCheck: new Date(),
  };

  constructor(config: EventLogConfig) {
    this.dbPath = config.dbPath;
    this.circuitBreaker = config.circuitBreaker || {
      windowMinutes: 60,
      threshold: 1000000, // 1M credits default
      enabled: true,
    };
    this.pythonEngine = new PythonEngine();
    this.initializeDatabase();
  }

  /**
   * Initialize DuckDB database and schema via Python
   */
  private async initializeDatabase(): Promise<void> {
    try {
      const scriptPath = join(process.cwd(), 'tools/observability/event_log.py');
      const resultSchema = z.object({
        success: z.boolean(),
        error: z.string().optional(),
      });

      await this.pythonEngine.runScript(
        scriptPath,
        {
          operation: 'init',
          'db-path': this.dbPath,
        },
        resultSchema
      );

      logger.info('Event log database initialized', { dbPath: this.dbPath });
    } catch (error) {
      logger.error('Failed to initialize event log database', error as Error, {
        dbPath: this.dbPath,
      });
      // Don't throw - allow service to continue with degraded functionality
    }
  }

  /**
   * Log an API call event
   */
  async logEvent(entry: Omit<EventLogEntry, 'id'>): Promise<void> {
    // Check circuit breaker before logging
    if (this.circuitBreaker.enabled) {
      const shouldBlock = await this.checkCircuitBreaker();
      if (shouldBlock) {
        logger.warn('Circuit breaker tripped - blocking API call', {
          apiName: entry.api_name,
          endpoint: entry.endpoint,
          trippedAt: this.circuitState.trippedAt,
        });
        throw new AppError(
          `Circuit breaker tripped: Credits exceeded threshold (${this.circuitBreaker.threshold}) in last ${this.circuitBreaker.windowMinutes} minutes`,
          'CIRCUIT_BREAKER_TRIPPED',
          503,
          {
            threshold: this.circuitBreaker.threshold,
            windowMinutes: this.circuitBreaker.windowMinutes,
            entry,
          }
        );
      }
    }

    try {
      const id = crypto.randomUUID();
      const validated = EventLogEntrySchema.parse({ ...entry, id });

      const scriptPath = join(process.cwd(), 'tools/observability/event_log.py');
      const resultSchema = z.object({
        success: z.boolean(),
        error: z.string().optional(),
      });

      await this.pythonEngine.runScript(
        scriptPath,
        {
          operation: 'log',
          'db-path': this.dbPath,
          event: JSON.stringify({
            id: validated.id,
            timestamp: validated.timestamp.toISOString(),
            api_name: validated.api_name,
            endpoint: validated.endpoint,
            method: validated.method,
            status_code: validated.status_code,
            success: validated.success,
            latency_ms: validated.latency_ms,
            credits_cost: validated.credits_cost,
            run_id: validated.run_id,
            error_message: validated.error_message,
            metadata: validated.metadata,
          }),
        },
        resultSchema
      );

      logger.debug('Event logged', {
        apiName: validated.api_name,
        endpoint: validated.endpoint,
        creditsCost: validated.credits_cost,
      });
    } catch (error) {
      logger.error('Failed to log event', error as Error, {
        apiName: entry.api_name,
        endpoint: entry.endpoint,
      });
      // Don't throw - event logging should not break application flow
    }
  }

  /**
   * Check circuit breaker status
   */
  private async checkCircuitBreaker(): Promise<boolean> {
    if (!this.circuitBreaker.enabled) {
      return false;
    }

    try {
      const windowStart = DateTime.now()
        .minus({ minutes: this.circuitBreaker.windowMinutes })
        .toJSDate();

      const scriptPath = join(process.cwd(), 'tools/observability/event_log.py');
      const resultSchema = z.object({
        total_credits: z.number(),
      });

      const result = await this.pythonEngine.runScript(
        scriptPath,
        {
          operation: 'get_credits_in_window',
          'db-path': this.dbPath,
          'window-start': windowStart.toISOString(),
        },
        resultSchema
      );

      const totalCredits = result.total_credits || 0;

      if (totalCredits >= this.circuitBreaker.threshold) {
        if (!this.circuitState.tripped) {
          this.circuitState.tripped = true;
          this.circuitState.trippedAt = new Date();
          logger.error('Circuit breaker TRIPPED', {
            totalCredits,
            threshold: this.circuitBreaker.threshold,
            windowMinutes: this.circuitBreaker.windowMinutes,
          });
        }
        return true;
      } else {
        // Reset if below threshold
        if (this.circuitState.tripped) {
          logger.info('Circuit breaker RESET', {
            totalCredits,
            threshold: this.circuitBreaker.threshold,
          });
          this.circuitState.tripped = false;
          this.circuitState.trippedAt = null;
        }
        return false;
      }
    } catch (error) {
      logger.error('Failed to check circuit breaker', error as Error);
      // On error, don't block (fail open)
      return false;
    }
  }

  /**
   * Get credits spent in last N minutes
   */
  async getCreditsInWindow(minutes: number): Promise<number> {
    try {
      const windowStart = DateTime.now().minus({ minutes }).toJSDate();

      const scriptPath = join(process.cwd(), 'tools/observability/event_log.py');
      const resultSchema = z.object({
        total_credits: z.number(),
      });

      const result = await this.pythonEngine.runScript(
        scriptPath,
        {
          operation: 'get_credits_in_window',
          'db-path': this.dbPath,
          'window-start': windowStart.toISOString(),
        },
        resultSchema
      );

      return result.total_credits || 0;
    } catch (error) {
      logger.error('Failed to get credits in window', error as Error);
      return 0;
    }
  }

  /**
   * Get event log statistics
   */
  async getStats(
    apiName?: string,
    startDate?: Date,
    endDate?: Date
  ): Promise<{
    totalEvents: number;
    totalCredits: number;
    successRate: number;
    avgLatency: number;
    errorCount: number;
  }> {
    try {
      const scriptPath = join(process.cwd(), 'tools/observability/event_log.py');
      const resultSchema = z.object({
        total_events: z.number(),
        total_credits: z.number(),
        success_rate: z.number(),
        avg_latency: z.number(),
        error_count: z.number(),
      });

      const result = await this.pythonEngine.runScript(
        scriptPath,
        {
          operation: 'get_stats',
          'db-path': this.dbPath,
          'api-name': apiName,
          'start-date': startDate?.toISOString(),
          'end-date': endDate?.toISOString(),
        },
        resultSchema
      );

      return {
        totalEvents: result.total_events || 0,
        totalCredits: result.total_credits || 0,
        successRate: (result.success_rate || 0) * 100,
        avgLatency: result.avg_latency || 0,
        errorCount: result.error_count || 0,
      };
    } catch (error) {
      logger.error('Failed to get event log stats', error as Error);
      throw error;
    }
  }

  /**
   * Check if circuit breaker is tripped
   */
  isCircuitBreakerTripped(): boolean {
    return this.circuitState.tripped;
  }

  /**
   * Close database connection
   */
  async close(): Promise<void> {
    // Python engine handles cleanup automatically
    logger.info('Event log service closed');
  }
}

/**
 * Singleton instance
 */
let eventLogInstance: EventLogService | null = null;

/**
 * Get or create the singleton EventLogService instance
 */
export function getEventLogService(config?: EventLogConfig): EventLogService {
  if (!eventLogInstance) {
    if (!config) {
      throw new ConfigurationError(
        'EventLogService config required for first initialization',
        'EventLogService'
      );
    }
    eventLogInstance = new EventLogService(config);
  }
  return eventLogInstance;
}
